Support multiple Redis clients

machi1990 opened the pull request:
This is into two separate commits as the first one only add some additional configuration options.

The second commit is the "real" subject of this PR.

Fixes https://github.com/quarkusio/quarkus/issues/11558

machi1990 committed:
feat(redis): add the possibility to configure "maxWaitingHandlers" and "maxNestedArrays" options

machi1990 committed:
feat(redis): support for multiple redis clients

Fixes https://github.com/quarkusio/quarkus/issues/11558

gsmet commented on extensions/redis-client/deployment/src/main/java/io/quarkus/redis/client/deployment/RedisClientProcessor.java:
```diff
@@ -39,4 +72,101 @@ HealthBuildItem addHealthCheck(RedisBuildTimeConfig buildTimeConfig) {
+            syntheticBeans.produce(createReactiveSyntheticBean(recorder, clientName));
+            syntheticBeans.produce(createMutinyRedisApiSyntheticBean(recorder, clientName));
+            syntheticBeans.produce(createMutinyRedisClientSyntheticBean(recorder, clientName));
+            syntheticBeans.produce(createVertxRedisAPISyntheticBean(recorder, clientName));
```
The one two lines above is called `Api` not `API`. Could we be consistent?

machi1990 replied:
Good point, thanks for spotting this. I had this mixed up all over the codebase and I've applied some clean up.

machi1990 committed:
feat(redis): make the RedisConfiguration matches the upstream `RedisOptions`

This will make it easier to tweak things like redis connection hosts to match the expected endpoint
by the upstream library thus avoid manually reconstructing the connection string. See
https://github.com/vert-x3/vertx-redis-client/blob/e9e3c0db349d662e676c04235841c68488d6f6d2/src/main/java/io/vertx/redis/client/RedisOptions.java
for more details about the available options. The global default password for cluster and sentinel
mode will be added once we have the next release of vertx-redis. `NetClientOptions` can be added now
but I considered it to be of less importance and we can have separate PR to configure it once users
asks for it.

Fixes https://github.com/quarkusio/quarkus/issues/11986

vitor-cardoso-tdx commented:
Looks good , just a question can we port this improvement to 1.7.x some next version because right now I cannot use the sentinel mode: https://github.com/quarkusio/quarkus/issues/11986

gsmet commented:
@machi1990 I think we need a separate PR for the Sentinel fix targeting 1.8 as it looks like something we will want in 1.8.1.

machi1990 commented:
> @machi1990 I think we need a separate PR for the Sentinel fix targeting 1.8 as it looks like something we will want in 1.8.1.

To fix the issue, I had to refactor the redis configuration a little. Is that something that's desirable for 1.8.1?

machi1990 commented:
> @machi1990 yeah, probably not a good idea.

Great. I'll be inclined for the fix to go in with this PR at a later release.

gsmet commented on docs/src/main/asciidoc/redis.adoc:
```diff
@@ -544,6 +544,40 @@ So when you access the `/health/ready` endpoint of your application you will hav
 
+== Multiple Redis Clients
+
+Redis allows you to configure multiple clients.
```
```suggestion
The Redis extension allows you to configure multiple clients.
```
maybe?

gsmet commented on docs/src/main/asciidoc/redis.adoc:
```diff
@@ -544,6 +544,40 @@ So when you access the `/health/ready` endpoint of your application you will hav
+
+=== Named Redis client Injection
+
+When using multiple clients, each `RedisClient`, you can select the client to inject using the `io.quarkus.redis.client.RedisClientName` qualifier.
```
There's something weird with this sentence, isn't it? The `each RedisClient` part sounds weird.

machi1990 replied:
Yeap. That part should not be there.

gsmet commented on extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/RedisClientName.java:
```diff
@@ -0,0 +1,41 @@
+import javax.inject.Qualifier;
+
+/**
+ * Marker annotation to select redis client
```
```suggestion
 * Marker annotation to select the Redis client.
```

gsmet commented on extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/RedisClientName.java:
```diff
@@ -0,0 +1,41 @@
+ * Marker annotation to select redis client
+ * Use name parameter to select it
+ *
+ * For example, if a redis connection is configured like so in {@code application.properties}:
```
```suggestion
 * For example, if a Redis connection is configured like so in {@code application.properties}:
```

gsmet commented on extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/RedisClientName.java:
```diff
@@ -0,0 +1,41 @@
+
+/**
+ * Marker annotation to select redis client
+ * Use name parameter to select it
```
Let's just drop this sentence, it's just noise.

gsmet commented on extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/runtime/RedisClientUtil.java:
```diff
@@ -0,0 +1,63 @@
+import io.vertx.redis.client.RedisOptions;
+
+public class RedisClientUtil {
+    public static final String DEFAULT_CLIENT = "__redis_default_client__";
```
In the other extensions, we call it `<default>`.

machi1990 replied:
Sure, let's be consistent here too. I'll change it.

gsmet commented on extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/runtime/RedisConfig.java:
```diff
@@ -1,77 +1,136 @@
 
     /**
-     * The redis hosts
+     * Configures additional redis client connections.
```
```suggestion
     * Configures additional Redis client connections.
```

gsmet commented on extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/runtime/RedisConfig.java:
```diff
@@ -1,77 +1,136 @@
+     * </pre>
+     * <p>
+     * And then use annotations above the instances of RedisClient to indicate which instance we are going to use
+     * <p>
```
Was it copy/pasted from elsewhere? All this Javadoc block could use some rework.

machi1990 replied:
I'll push some updates. Let me know what you think.

gsmet reviewed (changes requested):
Looks mostly good. I added some comments.

machi1990 commented:
@gsmet thanks for the review. I've addressed your comments. Let me know what you think.

gsmet commented on extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/RedisClientName.java:
```diff
@@ -0,0 +1,40 @@
+import javax.inject.Qualifier;
+
+/**
+ * Marker annotation to select the redis client
```
```suggestion
 * Marker annotation to select the Redis client.
```
(it's minor but this is public API so let's get it right)

gsmet commented on extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/RedisClientName.java:
```diff
@@ -0,0 +1,40 @@
+    /**
+     * Specify the cluster name of the connection.
+     *
+     * @return the value
```
Let's drop this.

gsmet commented on extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/RedisClientName.java:
```diff
@@ -0,0 +1,40 @@
+@Qualifier
+public @interface RedisClientName {
+    /**
+     * Specify the cluster name of the connection.
```
```suggestion
     * The Redis client name.
```

gsmet commented on extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/runtime/RedisConfig.java:
```diff
@@ -1,77 +1,140 @@
-     * The redis hosts
+     * Configures additional Redis client connections.
+     * <p>
+     * Each client have a unique identifier which must be identified to select the right connection.
```
```suggestion
     * Each client has a unique identifier which must be identified to select the right connection.
```

gsmet commented on extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/runtime/RedisConfig.java:
```diff
@@ -1,77 +1,140 @@
+     * quarkus.redis.client2.hosts = redis://localhost:6380
+     * </pre>
+     * <p>
+     * And then use the {@link RedisClientName} to select the {@link RedisClient} or
+     * {@link io.quarkus.redis.client.reactive.ReactiveRedisClient}
+     * we are going to use.
```
```suggestion
     * And then use the {@link RedisClientName} annotation to select the {@link RedisClient} or
     * {@link io.quarkus.redis.client.reactive.ReactiveRedisClient}.
```

gsmet commented on extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/runtime/RedisAPIProducer.java:
```diff
@@ -1,137 +1,55 @@
-import io.vertx.redis.client.RedisClientType;
 import io.vertx.redis.client.RedisOptions;
 
 class RedisAPIProducer {
```
Is there a reason why we are not being consistent and use the synthetic bean pattern for this too?

machi1990 replied:
You mean drop this build item  https://github.com/quarkusio/quarkus/blob/6a7ba55365abf623a196fe106800c490c88fe6ac/extensions/redis-client/deployment/src/main/java/io/quarkus/redis/client/deployment/RedisClientProcessor.java#L59-L64  and use a synthetic bean instead? 

If so, I saw no need of the synthetic bean for this one.

gsmet replied:
OK, we can live with it I suppose.

gsmet reviewed (changes requested):
Added some additional suggestions and a question.

machi1990 committed:
docs(redis): various documentation adjustements in the redis extension

Co-authored-by: Guillaume Smet <guillaume.smet@gmail.com>

machi1990 commented:
@gsmet thanks for the suggested fixes which I've applied. I've also answered https://github.com/quarkusio/quarkus/pull/11908#discussion_r490144731, let me know what you think.

machi1990 commented:
Merged. @gsmet thanks for the review.

gsmet commented:
Thanks for your work, it's a nice addition!

geoand commented:
This is a breaking change, right?

machi1990 commented:
> This is a breaking change, right?

Right. See https://github.com/quarkusio/quarkus/pull/11908#issuecomment-689527908.

gsmet commented:
Well, there's a difference between refactoring the config and breaking the config :).

What's the rationale for removing the `ssl` bit and include everything in the `host`? Also, could we be a bit better on that if we really need it i.e. support the old config for a while but raise a warning?

machi1990 commented:
> Well, there's a difference between refactoring the config and breaking the config :).

Right. 

> What's the rationale for removing the `ssl` bit and include everything in the `host`? Also, could we be a bit better on that if we really need it i.e. support the old config for a while but raise a warning?

Breaking up the connection string into several parts worked nice with a standalone redis client type but not so well with cluster or sentinel.  There was no point of exposing the property and doing construction of the connection string in the extension while the user can supply the connection string.  As for `ssl`, the scheme would have to `rediss`.   

> Also, could we be a bit better on that if we really need it i.e. support the old config for a while but raise a warning?

I agree and I would have done it this way really.

machi1990 commented:
> This is a breaking change, right?

Quick update regarding this: In the coming days, I'll open a followup PR to provide backward compatibility for the removed configuration properties for at least one release. This will give users ample time to move to the new way of configuring things.

geoand commented:
Thanks!
